### PR TITLE
feat: Add metadata fields to core entities 

### LINKED ISSUE #1610: Add meta data fields to core entities
# Overview 

Larger teams often require the ability to add their own meta-data to the following entities:

- Flags
- Segments
- Environments

Examples of the meta data that they might add:

- JIRA ticket code (e.g. `FLAGSMITH-152`) to a Flag
- URL for Infra overview for an Environment
- Link to Amplitude for a Segment

These data could be optional or required. For example, engineers should be required to add a JIRA ticket against a new Flag when the flag is created

# Implementation

Administrators should be able to define fields that they want associated with an entity, and whether it is required or optional.

To begin with the administrator should be able to define with the field is single line or multi-line. For v1, just allow any text. 

gz#40

### PR DESCRIPTION
**Pull request by novakzaballa**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes
FE changes:
- Add and create metadata to Environments, segments, and Flags
- The feature was wrapped inside the 'enable_metadata' flag
BE changes:
- Support metadata for segments, and Flags
- Support required metadata model fields for segments, and Flags

## How did you test this code?

- Go to manage
- Click the Metadata tab.
- Create Metadata.
- Edit metadata and enable any entity

### CODE REVIEW

**Comment by matthewelwell** [comment]
@novakzaballa the main thing I have found so far is that there is no feedback when you create a new metadata item. The side panel doesn't close and there is no reloading of the list. It does also seem that perhaps it's not always reading / writing to the same organisation but I'm not able to consistently reproduce that. 

https://github.com/Flagsmith/flagsmith/assets/14089968/08e16122-ae02-4ad3-a6a4-1e430a0fbfd9

**Comment by novakzaballa** [comment]
@matthewelwell 
> @novakzaballa the main thing I have found so far is that there is no feedback when you create a new metadata item. The side panel doesn't close and there is no reloading of the list. 
> 
This has already been corrected.

> It does also seem that perhaps it's not always reading / writing to the same organisation but I'm not able to consistently reproduce that.
>

I'll take a look to see if I can reproduce what you mentioned.

File: frontend/web/components/modals/CreateMetadata.tsx [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+          className: 'full-width',
+          name: 'Name',
+        }}
+        value={name}
+        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
+          setName(Utils.safeParseEventValue(event))
+        }}
+        type='text'
+        id='metadata-name'
+        placeholder='FL-124'
```
Comment (matthewelwell) [inline frontend/web/components/modals/CreateMetadata.tsx]:
```suggestion
        placeholder='e.g. JIRA Ticket Number'
```

File: frontend/web/components/modals/CreateMetadata.tsx [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+        inputProps={{
+          className: 'full-width',
+          name: 'metadata-desc',
+        }}
+        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
+          setDescription(Utils.safeParseEventValue(event))
+        }}
+        type='text'
+        title={'Description (optional)'}
+        placeholder='This is a metadata description'
```
Comment (matthewelwell) [inline frontend/web/components/modals/CreateMetadata.tsx]:
```suggestion
        placeholder='e.g. \'The JIRA Ticket Number associated with this flag\''
```

File: frontend/web/components/modals/CreateFlag.js [hunk 3]
```diff
@@ -460,6 +477,8 @@ const CreateFlag = class extends Component {
     } = this.state
     const FEATURE_ID_MAXLENGTH = Constants.forms.maxLength.FEATURE_ID
 
+    // const metadataList = this.getApprovals(metadata, this.state.metadata || [])
```
Comment (matthewelwell) [inline frontend/web/components/modals/CreateFlag.js]: Can we remove this?

**Review by matthewelwell — CHANGES_REQUESTED** [review]
Some more issues I've found in manual testing: 

image

This UI is kind of odd, especially since these toggles aren't even useable. Can we just use some symbols here to determine if it's enabled / required and add an edit option at the end of the row, next to 'Remove' ? 

https://github.com/Flagsmith/flagsmith/assets/14089968/4f5188af-68cb-4744-ac38-4c8c104af3ce

I can't seem to make a metadata field required for any entity. 

image

The UI lets me configure the metadata as required (although it doesn't persist it, see above) even if the field is not enabled for that entity type.

File: frontend/web/components/modals/CreateFlag.js (hunk 3 above)
Comment (novakzaballa) [reply frontend/web/components/modals/CreateFlag.js]: Removed.

File: frontend/web/components/modals/CreateMetadata.tsx (hunk 2 above)
Comment (novakzaballa) [reply frontend/web/components/modals/CreateMetadata.tsx]: Done.

File: frontend/web/components/modals/CreateMetadata.tsx (hunk 1 above)
Comment (novakzaballa) [reply frontend/web/components/modals/CreateMetadata.tsx]: Done.

File: api/segments/tests/test_views.py
Comment (matthewelwell) [inline api/segments/tests/test_views.py:1]: 2 things here: 
1. This isn't actually testing creating a segment, it's still just creating an environment. 
2. This test module should live at `api/tests/unit/segments/test_unit_segments_views.py`

File: api/features/tests/test_views.py
Comment (matthewelwell) [inline api/features/tests/test_views.py:1]: 2 things here: 
1. This isn't actually testing creating a feature, it's still just creating an environment. 
2. This test module should live at `api/tests/unit/features/test_unit_features_views.py`

**Review by matthewelwell — CHANGES_REQUESTED** [review]
I haven't been through the rest of the PR, but just finishing my review here as we need to decide on whether we should include the addition of Feature and Segment metadata as part of this PR or another PR. 

I have a feeling that once you update the tests that I have added comments to, there will be a substantial amount of additional work (primarily serializers I think) to get that functionality working which may blow out this PR.

File: api/features/serializers.py [hunk 4]
```diff
@@ -235,6 +236,29 @@ def get_num_identity_overrides(self, instance) -> typing.Optional[int]:
...
+    metadata = MetadataSerializer(required=False, many=True)
+
+    class Meta(ListCreateFeatureSerializer.Meta):
+        fields = ListCreateFeatureSerializer.Meta.fields + ("metadata",)
+
+    def get_project(self, validated_data: dict = None) -> Project:
+        view = self.context.get("view")
+
+        if view and "project_pk" in view.kwargs:
+            project_pk = view.kwargs["project_pk"]
```
Comment (matthewelwell) [inline api/features/serializers.py:261]: Have we verified that all uses of this serializer are nested beneath projects?

File: api/tests/unit/features/test_unit_features_views.py [hunk 5]
```diff
@@ -1273,6 +1273,42 @@ def test_cannot_update_feature_of_a_feature_state(
     )
 
 
+@pytest.mark.parametrize(
+    "client",
+    [lazy_fixture("admin_master_api_key_client"), lazy_fixture("admin_client")],
+)
+def test_create_feature_with_required_metadata_returns_201(
```
Comment (matthewelwell) [inline api/tests/unit/features/test_unit_features_views.py:2226]: Can we add a similar test for segments?

File: frontend/common/types/responses.ts [hunk 6]
```diff
@@ -376,6 +376,8 @@ export type Res = {
   rolePermissionGroup: { id: string }
   getSubscriptionMetadata: { id: string }
   environment: Environment
+  metadataModelField: { id: string }
+  metadata: { id: string }
```
Comment (matthewelwell) [inline frontend/common/types/responses.ts]: Maybe I don't quite understand the implementation here, but it feels like there should be more fields defined here?

File: frontend/web/components/modals/CreateFlag.js (hunk 3 above)
Comment (matthewelwell) [reply frontend/web/components/modals/CreateFlag.js]: Still seems to be there :/

**Review by matthewelwell — CHANGES_REQUESTED** [review]
Added some comments on the code below, but also some general feedback from functional testing. 

1. I think this would look nicer without the 'Enabled' / 'Disabled' / 'Edit Metadata' text? 

image

2. I still can't persist a required metadata field. I toggle the switcher to make it 'required' for environment, hit save then open up the metadata again and it's not required. 
3. We should add a confirm step to deleting a metadata item. 
4. I'm not sure why we have a '*' here: 

image

5. I can't add metadata to an entity. Clicking on the metadata item doesn't do anything. This is the case for both features and environments. 

https://github.com/Flagsmith/flagsmith/assets/14089968/1e2b61ee-2b27-4e47-be63-865c004994ff

6. The search input doesn't seem to be hooked up to anything: 

image

7. I understood that the Metadata items were supposed to live at the organisation level, not the project. They can be _required_ for a given entity type for a project, but they should live at the organisation level. Correct me if I'm wrong here @gagantrivedi.

**Comment by dabeeeenster** [comment]
Should we make metadata definitions not-updateable/write once? What happens if you try and change a `string` for an `int` etc? It just feels dangerous/not applicable?

**Comment by novakzaballa** [comment]
> Added some comments on the code below, but also some general feedback from functional testing.
>
> 1. I think this would look nicer without the 'Enabled' / 'Disabled' / 'Edit Metadata' text?

Yes, texts were deleted.


> 3. We should add a confirm step to deleting a metadata item.

Confirm modal added.

> 4. I'm not sure why we have a '*' here:

Deleted


> 6. The search input doesn't seem to be hooked up to anything:

Corrected

The other 2 items are still in progress

File: frontend/common/types/responses.ts (hunk 6 above)
Comment (novakzaballa) [reply frontend/common/types/responses.ts]: Right, Done.

**Comment by novakzaballa** [comment]
> Still seems to be there :/

Corrected.

File: frontend/common/constants.ts [hunk 7]
```diff
@@ -212,6 +212,12 @@ export default {
     primary: '#6837fc',
     white: '#ffffff',
   },
+  contentTypes: {
+    environment: 30,
```
Comment (kyle-ssg) [inline frontend/common/constants.ts:217]: Are these hardcoded ids ? Not clear what this is. If they are ids, does this differ per environment or are they somehow fixed.

File: frontend/common/constants.ts [hunk 8]
```diff
@@ -466,6 +472,7 @@ export default {
       'Set different values for your feature based on what segments users are in.',
     TAGS_DESCRIPTION:
       'Organise your flags with tags, tagging your features as "<strong>protected</strong>" will prevent them from accidentally being deleted.',
+    TOOLTIP_METADATA_DESCRIPTION: 'Add metadata in your',
```
Comment (kyle-ssg) [inline frontend/common/constants.ts:476]: Doesn't look complete

**Comment by kyle-ssg** [comment]
> image

I think this needs to go to design or at least more closely use the design system, e.g. making these column items checkboxes rather than the current ui and a less wordy edit cta.

Do we explain what "Metadata" is where the user can create it ? Maybe a subtitle and a link to docs would be useful here.

File: frontend/common/constants.ts (hunk 7 above)
Comment (novakzaballa) [reply frontend/common/constants.ts:217]: These are the IDs of the content types of those 4 entities, they do not change.
Screen Shot 2023-12-14 at 16 32 38

File: frontend/common/constants.ts (hunk 8 above)
Comment (novakzaballa) [reply frontend/common/constants.ts:476]: It is not incomplete, it is used like this `tooltip={${Constants.strings.TOOLTIP_METADATA_DESCRIPTION} segments}`

**Comment by novakzaballa** [comment]
> 
> 
> I think this needs to go to design or at least more closely use the design system, e.g. making these column items checkboxes rather than the current ui and a less wordy edit cta.
> 

I made some changes.
> Do we explain what "Metadata" is where the user can create it ? Maybe a subtitle and a link to docs would be useful here.

I put a metadata subtitle, in the sections "managing features", "managing segments" and "environment settings"  inside the docs.

File: frontend/web/components/MetadataSelect.tsx [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      const search = filter.toLowerCase()
+      if (!search) return true
+      return `${v.name}`.toLowerCase().includes(search)
+    })
+  return (
+    <InlineModal
+      title='Metadata'
+      isOpen={isOpen}
+      onClose={onToggle}
+      className='inline-modal--tags'
```
Comment (kyle-ssg) [inline frontend/web/components/MetadataSelect.tsx]: Can this be replaced with inline-modal--sm

File: frontend/web/components/MetadataSelect.tsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    >
+      <Input
+        disabled={disabled}
+        value={filter}
+        onChange={(e: InputEvent) => setFilter(Utils.safeParseEventValue(e))}
+        className='full-width mb-2'
+        placeholder='Type or choose a Metadata'
+      />
+      <div style={{ maxHeight: 200, overflowY: 'auto' }}>
```
Comment (kyle-ssg) [inline frontend/web/components/MetadataSelect.tsx]: Can the inline style be replaced with className="inline-modal__list"

File: frontend/web/components/MetadataSelect.tsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        {metadataList &&
```
Comment (kyle-ssg) [inline frontend/web/components/MetadataSelect.tsx]: What if there are none? Should display a default message

```
metadataList?.length? () : (<div className="text-center">No items</div>)
```

File: frontend/web/components/MetadataSelect.tsx [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+                  if (isRemove && onRemove) {
+                    onRemove(v.id)
+                  } else if (!isRemove && onAdd) {
+                    onAdd(v.id)
+                  }
+                }}
+                space
+              >
+                <Flex
+                  className={value?.includes(v.id) ? 'font-weight-bold' : ''}
```
Comment (kyle-ssg) [inline frontend/web/components/MetadataSelect.tsx]: If possible can we rename and reuse TableFilterItem.tsx? It does the exact same as this.

File: frontend/web/components/MyMetadataSelect.tsx [hunk 13]
```diff
@@ -0,0 +1,37 @@
+import { FC, useEffect, useState } from 'react'
```
Comment (kyle-ssg) [inline frontend/web/components/MyMetadataSelect.tsx:1]: Is there a reason why it's called MyMetadata, that would indicate to me it's specific to the user but I think this is for the org?

File: frontend/web/components/modals/CreateFlag.js [hunk 14]
```diff
@@ -531,6 +576,73 @@ const CreateFlag = class extends Component {
...
+                        style={{ marginBottom: 4, marginTop: 4 }}
+                      >
+                        <Tooltip
+                          title={
+                            <Row>
+                              <span className='font-weight-bold'>{m.name}</span>
+                              <span className='chip-icon ion'>
+                                <IonIcon
+                                  icon={closeIcon}
+                                  style={{ fontSize: '13px' }}
```
Comment (kyle-ssg) [inline frontend/web/components/modals/CreateFlag.js:602]: className="fs-caption"

File: frontend/web/components/pages/ProjectSettingsPage.js [hunk 15]
```diff
@@ -498,6 +603,315 @@ const ProjectSettingsPage = class extends Component {
...
+                                  </Flex>
+                                  <div
+                                    className='table-column'
+                                    style={{
+                                      alignItems: 'center',
+                                      display: 'flex',
+                                      width: '185px',
+                                    }}
+                                  >
+                                    <div className='table-column'>
```
Comment (kyle-ssg) [inline frontend/web/components/pages/ProjectSettingsPage.js]: Markup here doesn't lookl correct, we are nesting table-column elements. Please lets use classnames where possible too.

```
                                    <div className='table-column d-flex text-center'>

```

File: frontend/web/components/pages/ProjectSettingsPage.js [hunk 16]
```diff
@@ -498,6 +603,315 @@ const ProjectSettingsPage = class extends Component {
...
+                                      {metadata.content_type_fields.find(
```
Comment (kyle-ssg) [inline frontend/web/components/pages/ProjectSettingsPage.js:679]: Hard to read what the following functions are doing, perhaps separate these into variables above the markup ?

File: frontend/web/components/pages/ProjectSettingsPage.js [hunk 17]
```diff
@@ -498,6 +603,315 @@ const ProjectSettingsPage = class extends Component {
...
+                                  </div>
+                                  <div
+                                    className='table-column'
+                                    style={{ width: '150px' }}
+                                  >
+                                    <div
+                                      className='table-column'
+                                      style={{ width: '150px' }}
+                                    >
+                                      {metadata.content_type_fields.find(
```
Comment (kyle-ssg) [inline frontend/web/components/pages/ProjectSettingsPage.js:744]: Same here

File: frontend/web/components/pages/ProjectSettingsPage.js [hunk 18]
```diff
@@ -498,6 +603,315 @@ const ProjectSettingsPage = class extends Component {
...
+                                      style={{ width: '150px' }}
+                                      onClick={() => {
+                                        this.editMetadata(
+                                          metadata.id,
+                                          metadata.content_type_fields,
+                                        )
+                                      }}
+                                    >
+                                      {metadata.content_type_fields.find(
+                                        (m) =>
```
Comment (kyle-ssg) [inline frontend/web/components/pages/ProjectSettingsPage.js:821]: And here

File: frontend/web/components/MetadataSelect.tsx (hunk 9 above)
Comment (novakzaballa) [reply frontend/web/components/MetadataSelect.tsx]: Done

File: frontend/web/components/MetadataSelect.tsx (hunk 10 above)
Comment (novakzaballa) [reply frontend/web/components/MetadataSelect.tsx]: Done

File: frontend/web/components/MetadataSelect.tsx (hunk 11 above)
Comment (novakzaballa) [reply frontend/web/components/MetadataSelect.tsx]: Done

File: frontend/web/components/modals/CreateFlag.js (hunk 14 above)
Comment (novakzaballa) [reply frontend/web/components/modals/CreateFlag.js:602]: Done

File: frontend/web/components/pages/ProjectSettingsPage.js (hunk 15 above)
Comment (novakzaballa) [reply frontend/web/components/pages/ProjectSettingsPage.js]: Done

**Pull request closed without merging**